File: Backend/src/main/java/com/luckyseven/backend/domain/settlements/dao/SettlementSpecification.java
package com.luckyseven.backend.domain.settlements.dao;

import com.luckyseven.backend.domain.settlements.dto.SettlementSearchCondition;
import com.luckyseven.backend.domain.settlements.entity.Settlement;
import org.springframework.data.jpa.domain.Specification;

public class SettlementSpecification {

  public static Specification<Settlement> hasTeamId(Long teamId) {
    return (root, query, cb) -> teamId == null ? null :
        cb.equal(root.get("expense").get("team").get("id"), teamId);
  }

  public static Specification<Settlement> hasPayerId(Long payerId) {
    return (root, query, criteriaBuilder) -> payerId == null ? null
        : criteriaBuilder.equal(root.get("payer").get("id"), payerId);
  }

  public static Specification<Settlement> hasSettlerId(Long settlerId) {
    return (root, query, criteriaBuilder) -> settlerId == null ? null
        : criteriaBuilder.equal(root.get("settler").get("id"), settlerId);
  }

  public static Specification<Settlement> hasExpenseId(Long expenseId) {
    return (root, query, criteriaBuilder) -> expenseId == null ? null
        : criteriaBuilder.equal(root.get("expense").get("id"), expenseId);
  }

  public static Specification<Settlement> isSettled(Boolean isSettled) {
    return (root, query, criteriaBuilder) -> isSettled == null ? null
        : criteriaBuilder.equal(root.get("isSettled"), isSettled);
  }

  public static Specification<Settlement> createSpecification(Long teamId,
      SettlementSearchCondition condition) {
    return Specification
        .where(SettlementSpecification.hasTeamId(teamId))
        .and(hasPayerId(condition.payerId))
        .and(hasSettlerId(condition.settlerId()))
        .and(hasExpenseId(condition.expenseId))
        .and(isSettled(condition.isSettled));
  }
}

File: Backend/src/main/kotlin/com/luckyseven/backend/domain/settlement/dao/SettlementSpecification.kt
package com.luckyseven.backend.domain.settlement.dao

import com.luckyseven.backend.domain.expense.entity.Expense
import com.luckyseven.backend.domain.member.entity.Member
import com.luckyseven.backend.domain.settlement.dto.SettlementSearchCondition
import com.luckyseven.backend.domain.settlement.entity.Settlement
import com.luckyseven.backend.domain.team.entity.Team
import org.springframework.data.jpa.domain.Specification

object SettlementSpecification {
    fun hasTeamId(teamId: Long?): Specification<Settlement> = Specification { root, _, cb ->
        if (teamId == null) null
        else cb.equal(root.get<Expense>("expense").get<Team>("team").get<Long>("id"), teamId)
    }

    fun hasPayerId(payerId: Long?): Specification<Settlement> = Specification { root, _, cb ->
        if (payerId == null) null
        else cb.equal(root.get<Member>("payer").get<Long>("id"), payerId)
    }

    fun hasSettlerId(settlerId: Long?): Specification<Settlement> = Specification { root, _, cb ->
        if (settlerId == null) null
        else cb.equal(root.get<Member>("settler").get<Long>("id"), settlerId)
    }

    fun hasExpenseId(expenseId: Long?): Specification<Settlement> = Specification { root, _, cb ->
        if (expenseId == null) null
        else cb.equal(root.get<Expense>("expense").get<Long>("id"), expenseId)
    }

    fun isSettled(isSettled: Boolean?): Specification<Settlement> = Specification { root, _, cb ->
        if (isSettled == null) null
        else cb.equal(root.get<Boolean>("isSettled"), isSettled)
    }

    fun createSpecification(
        teamId: Long?,
        condition: SettlementSearchCondition
    ): Specification<Settlement> = Specification
        .where(hasTeamId(teamId))
        .and(hasPayerId(condition.payerId))
        .and(hasSettlerId(condition.settlerId))
        .and(hasExpenseId(condition.expenseId))
        .and(isSettled(condition.isSettled))
}
